Add doc dir for Keptn 0.16 (#1228)
keptn/keptn.github.io: content/docs/0.16.x/integrations/custom_integration/index.md

If your Keptn-service is running in the same pod as the distributor (which we recommend), and receives events at the port `8080` and the path `/`, you will only need to set the `PUBSUB_TOPIC` environment variable.

</p>
</details>

**Bridge:**

After the application is deployed, its subscription can be changed directly in the Bridge.
For this, open Keptn Bridge, select a project, and go to the *Uniform* page.
Then select your service, and click the `Add subscription` button.

{{< popup_image
link="./assets/uniform_api.png"
caption="Add task subscription for your integration"
width="700px">}}

In this form, you can provide the information for the task subscription:

* *Task*: The task the integration should be fired on (e.g., `test` or `deployment`)
* *Task suffix*: The state of the task when the integration should be fired; select one of: `triggered`, `started`, of `finished`
* *Filter*: To restrict your integration to certain stages and services you can specify those using filters.

*Note:* multiple subscriptions can be added for the same service.

> Keptn stores this subscriptions information also if the integration is not running.
> By default, Keptn will still keep this information for 48h after the last contact with the integration.

Reviewer: ```suggestion
Keptn stores this subscriptions information even if the integration is not running.
By default, Keptn keeps this information for 48h after the last contact with the integration.
```

This value can be configured using the [Advanced Install](../../operate/advanced_install_options/) option `shipyardController.config.uniformIntegrationTTL`.

### Send a started event

> After receiving a `triggered` event for a particular task, your *Keptn-service* has to inform Keptn by sending an event of the type: 

Reviewer: ```suggestion
After receiving a `triggered` event for a particular task, your *Keptn-service* must inform Keptn by sending an event of the type: 
```

- `sh.keptn.event.[task].started`

The request body needs to follow the [CloudEvent specification](https://github.com/keptn/spec/blob/0.2.2/cloudevents.md) and the HTTP header attribute `Content-Type` has to be set to `application/cloudevents+json`.

**Send the event:**

> To send the event to Keptn, two ways are possible:

Reviewer: ```suggestion
The event can be sent to Keptn in either of the following ways:
```

1. Post it on the `v1/event` endpoint of Keptn
1. If the distributor is running as sidecar, post the event on `127.0.0.1:8081`

### Execute the functionality

The functionality of your *Keptn-service* depends on the capability you want to add to the continuous delivery or operational workflow. In many cases, the event payload -- containing meta-data such as the project, stage, or service name -- is first processed and then used to call the REST API of another tool.

### Send a finished event

After your *Keptn-service* has completed its functionality, it has to inform Keptn by sending an event of the type:

- `sh.keptn.event.[task].finished`

> The request body needs to follow the [CloudEvent specification](https://github.com/keptn/spec/blob/0.2.2/cloudevents.md) and the HTTP header attribute `Content-Type` has to be set to `application/cloudevents+json`.

Reviewer: ```suggestion
The request body must follow the [CloudEvent specification](https://github.com/keptn/spec/blob/0.2.2/cloudevents.md) and the HTTP header attribute `Content-Type` must be set to `application/cloudevents+json`.
```

* `status`: [succeeded, errored, unknown] - The status of the task execution.
* `result`: [pass, failed] - The result of a successful task execution.

```json
{
  "type": "sh.keptn.event.test.finished",
  "specversion": "1.0",
  "source": "https://github.com/keptn/keptn/jmeter-service",
  "id": "ggb878d3-03c0-4e8f-bc3f-454bc1b3d888",
  "time": "2019-06-07T07:02:15.64489Z",
  "contenttype": "application/json",
  "shkeptncontext": "08735340-6f9e-4b32-97ff-3b6c292bc509",
  "triggeredid" : "f2b878d3-03c0-4e8f-bc3f-454bc1b3d79d",
  "data": {
    "test": {
      "status": "succeeded",
      "result": "pass"
    },
    "project": "sockshop",
    "stage": "staging",
    "service": "carts",
    "labels": {
      "testId": "4711",
      "buildId": "build-17",
      "owner": "JohnDoe"
    }
  }
}
```

**Send the event:**

> To send the event to Keptn, two ways are possible:

Reviewer: ```suggestion
The event can be sent to Keptn, in either of the following ways:
```

              apiVersion: v1
              fieldPath: 'metadata.labels[''app.kubernetes.io/name'']'
        - name: K8S_POD_NAME
          valueFrom:
            fieldRef:
              apiVersion: v1
              fieldPath: metadata.name
        - name: K8S_NAMESPACE
          valueFrom:
            fieldRef:
              apiVersion: v1
              fieldPath: metadata.namespace
        - name: K8S_NODE_NAME
          valueFrom:
            fieldRef:
              apiVersion: v1
              fieldPath: spec.nodeName
```

To configure this distributor for your *Keptn-service*, two environment variables need to be adapted:

* `PUBSUB_RECIPIENT`: Defines the service address as specified in the Kubernetes service manifest (e.g., 127.0.0.1 or jmeter-service)
* `PUBSUB_TOPIC`: Defines the event type your *Keptn-service* is listening to (e.g. `sh.keptn.event.test.triggered`  or `sh.keptn.event.>`).

### Deploy Keptn-service and distributor

With a service and deployment manifest for your custom *Keptn-service* (`service.yaml`), you are ready to deploy both components in the K8s cluster where Keptn is installed:

```console
kubectl apply -f service.yaml -n keptn
```

## CloudEvents

> CloudEvents have to be sent with the HTTP header `Content-Type` set to `application/cloudevents+json`. For a detailed look into CloudEvents, please go the Keptn [CloudEvent specification](https://github.com/keptn/spec/blob/0.2.3/cloudevents.md).

Reviewer: ```suggestion
CloudEvents must be sent with the HTTP header `Content-Type` set to `application/cloudevents+json`. For a detailed look into CloudEvents, please go the Keptn [CloudEvent specification](https://github.com/keptn/spec/blob/0.2.3/cloudevents.md).
```

## Error Logging

> By default, the distributor will automatically extract error logs from received `sh.keptn.<task>.finished` events with `data.status=errored` and/or `data.result=fail`, that have been sent by your service. These error messages will then be forwarded to Keptn's Log Ingestion API.

Reviewer: ```suggestion
By default, the distributor automatically extracts error logs from received `sh.keptn.<task>.finished` events with `data.status=errored` and/or `data.result=fail`, that have been sent by your service. These error messages are then forwarded to Keptn's Log Ingestion API.
```

Additionally, for easier debugging of errors that occur either during the execution of a task of a sequence, or while performing any other operation, Keptn integration services can send error log events to the Keptn API via the distributor.
Examples for those events are listed below.

### Logging an error related task sequence execution:

> If the error log event should be associated to an execution of a specific task that has been triggered by a `sh.keptn.event.<task>.triggered` event, the following properties need to be set in order to correlate them to the correct task sequence execution:

Reviewer: ```suggestion
If the error log event is associated with the execution of a specific task that has been triggered by a `sh.keptn.event.<task>.triggered` event, the following properties must be set in order to correlate them to the correct task sequence execution:
```

- `shkeptncontext`: The context of the task sequence execution. Can be adapted from the received `sh.keptn.event.<task>.triggered` event
- `triggeredid`: The `id` of the received `sh.keptn.event.<task>.triggered` event
- `data.task`: The name of the executed task.
- `data.message`: The message you would like to log

**Example event payload**

```json
{
  "specversion": "1.0",
  "id": "c4d3a334-6cb9-4e8c-a372-7e0b45942f53",
  "source": "source-service",
  "type": "sh.keptn.log.error",
  "datacontenttype": "application/json",
  "data": {
    "message": "an unexpected error occurred during the execution of my task",
    "task": "deployment"
  },
  "triggeredid": "3f9640b6-1d2a-4f11-95f5-23259f1d82d6",
  "shkeptncontext": "a3e5f16d-8888-4720-82c7-6995062905c1",
  "shkeptnspecversion": "0.2.3"
}
```

### Logging an error that is not related to the execution of a task

> If the error log event should not be associated to an execution of a specific task, the properties `shkeptncontext`, `triggeredid`, and `data.task` are not required. In this case, an example payload would look as follows:

Reviewer: ```suggestion
If the error log event is not associated with an execution of a specific task, the properties `shkeptncontext`, `triggeredid`, and `data.task` are not required. In this case, an example payload would look as follows:
```
